samcli/lib/providers/sam_function_provider.py

```diff
@@ -387,43 +388,51 @@ def _parse_layer_info(
-                codeuri: Optional[str] = None
-
-                if resource_type == SamFunctionProvider.LAMBDA_LAYER:
-                    codeuri = SamFunctionProvider._extract_lambda_function_code(layer_properties, "Content")
-
-                if resource_type == SamFunctionProvider.SERVERLESS_LAYER:
-                    codeuri = SamFunctionProvider._extract_sam_function_codeuri(
-                        layer_logical_id, layer_properties, "ContentUri", ignore_code_extraction_warnings
-                    )
-
-                if codeuri and not use_raw_codeuri:
-                    LOG.debug("--base-dir is presented not, adjusting uri %s relative to %s", codeuri, stack.location)
-                    codeuri = SamLocalStackProvider.normalize_resource_path(stack.location, codeuri)
-
-                layers.append(
-                    LayerVersion(
-                        layer_logical_id,
-                        codeuri,
-                        compatible_runtimes,
-                        layer_resource.get("Metadata", None),
-                        stack_path=stack.stack_path,
-                    )
+                found_layer = SamFunctionProvider._locate_layer_from_ref(
+                    stack, layer, use_raw_codeuri, ignore_code_extraction_warnings
                 )
+                if found_layer:
+                    layers.append(found_layer)
 
         return layers
 
+    @staticmethod
+    def _locate_layer_from_ref(
+        stack: Stack, layer: Dict, use_raw_codeuri: bool = False, ignore_code_extraction_warnings: bool = False
+    ) -> Optional[LayerVersion]:
+        layer_logical_id = cast(str, layer.get("Ref"))
+        layer_resource = stack.resources.get(layer_logical_id)
+        if not layer_resource or layer_resource.get("Type", "") not in (
+            SamFunctionProvider.SERVERLESS_LAYER,
+            SamFunctionProvider.LAMBDA_LAYER,
+        ):
+            raise InvalidLayerReference()
+
+        layer_properties = layer_resource.get("Properties", {})
+        resource_type = layer_resource.get("Type")
+        codeuri: Optional[str] = None
+
+        if resource_type in [SamFunctionProvider.LAMBDA_LAYER, SamFunctionProvider.SERVERLESS_LAYER]:
+            code_property_key = SamBaseProvider.CODE_PROPERTY_KEYS[resource_type]
+            if SamBaseProvider._is_code_uri_s3(layer_properties.get(code_property_key)):
+                # Content can be a dictionary of S3 Bucket/Key or a S3 URI, neither of which are supported
+                if not ignore_code_extraction_warnings:
+                    SamFunctionProvider._warn_code_extraction(resource_type, layer_logical_id, code_property_key)
+                return None
+            codeuri = SamBaseProvider._extract_codeuri(layer_properties, code_property_key)
+
+        if codeuri and not use_raw_codeuri:
+            LOG.debug("--base-dir is presented not, adjusting uri %s relative to %s", codeuri, stack.location)
```

Comment:
There's some grammar issue on this sentence.

Reply:
good catch